### ISSUE TITLE
JAX/NumPy CTC loss can be slightly negative for confident alignments

### ISSUE THREAD

**Issue by Anu27n** [keras-team/keras#23684]
## Bug

Keras's JAX and NumPy CTC implementations are copied from Optax's log-space recursion. For a near-certain valid alignment, float32 rounding can make the per-sequence loss slightly **negative** (on the order of `1e-8`). A downstream `log(loss)` balancer then produces NaNs even though the data and model are finite.

This is the same class of bug as [google-deepmind/optax#1771](https://github.com/google-deepmind/optax/issues/1771). TensorFlow/PyTorch backends use their native CTC ops and are not affected.

## Reproduction (JAX backend)

```python
import numpy as np
import keras

labels = np.array([[1]], dtype="int32")
outputs = np.array([[[0.0, 17.0], [0.0, 17.0]]], dtype="float32")
loss = keras.ops.ctc_loss(labels, outputs, np.array([1]), np.array([2]), mask_index=0)
print(loss)  # can be slightly negative
```

Expected: nonnegative loss. Fix: clamp per-sequence loss at 0.

I have a fix with a regression test and will send a PR.

This PR was prepared with AI assistance. I reviewed the JAX/NumPy CTC implementations against the Optax reproduction and the added test.

**Comment by Anu27n** [comment]
Fix + tests: https://github.com/keras-team/keras/pull/23685

**Comment by Anu27n** [comment]
I opened this issue and the fix in https://github.com/keras-team/keras/pull/23685. The issue is currently assigned to @dhantule, so I will not treat it as mine.

If you want to take the change, I can close the PR. If the existing PR is useful, please reassign this issue to me so it matches Keras’s contributor policy.
